File: internal/trace/validate.go
package trace

import (
	"fmt"
	"strings"
)

// Severity of a validator diagnostic.
type Severity string

const (
	SevError   Severity = "error"
	SevWarning Severity = "warning"
)

// Diag is one validator finding.
type Diag struct {
	Check    string   `json:"check"`
	Severity Severity `json:"severity"`
	Event    int      `json:"event"` // event index, -1 for trace-level
	Message  string   `json:"message"`
}

func (d Diag) String() string {
	where := "trace"
	if d.Event >= 0 {
		where = fmt.Sprintf("event %d", d.Event)
	}
	return fmt.Sprintf("%s [%s] %s: %s", d.Check, d.Severity, where, d.Message)
}

// HasErrors reports whether any diagnostic is an error.
func HasErrors(ds []Diag) bool {
	for _, d := range ds {
		if d.Severity == SevError {
			return true
		}
	}
	return false
}

// Validate implements checks V1..V14 from TRACE_FORMAT.md 10.
//
// It MUST NOT panic on any input -- the JS twin of this function is the
// browser's trust boundary against arbitrary shared traces, and fuzz_test.go
// asserts the no-panic property here.
//
// V4 and V6 are the two that matter. Together they constitute a complete proof
// that the trace is a valid reversible delta log:
//
//	V4  replaying forward, every set.From deep-equals current state at set.At
//	V6  full forward replay then full backward replay returns the initial state
func Validate(t *Trace) (diags []Diag) {
	defer func() {
		if r := recover(); r != nil {
			diags = append(diags, Diag{"V0", SevError, -1,
				fmt.Sprintf("validator panicked: %v (this is a bug in Orrery)", r)})
		}
	}()

	add := func(check string, sev Severity, ev int, format string, a ...any) {
		diags = append(diags, Diag{check, sev, ev, fmt.Sprintf(format, a...)})
	}

	if t == nil {
		add("V1", SevError, -1, "trace is nil")
		return
	}

	// V1 -- version in range.
	if t.V < SupportedMin || t.V > SupportedMax {
		add("V1", SevError, -1, "version %d outside supported range [%d,%d]",
			t.V, SupportedMin, SupportedMax)
		return // nothing below is meaningful against an unknown version
	}

	// Declarations seen so far, and the aux flag per structure.
	decls := map[string]*Event{}
	aux := map[string]bool{}
	written := map[string]bool{}

	state := NewState(t)
	depth := 0
	prevG, seenG := 0, map[int]bool{}
	srcLines := 0
	if t.Meta.Source != nil {
		srcLines = strings.Count(t.Meta.Source.Text, "\n") + 1
	}

	// Snapshot of the initial hash for V6. The initial state is empty (no
	// structure exists before the first init), so this is the empty hash.
	initialHash := state.Hash()

	for i := range t.Events {
		e := &t.Events[i]

		// V7 -- group ids form contiguous runs.
		if e.G != 0 {
			if e.G != prevG && seenG[e.G] {
				add("V7", SevError, i, "group id %d reappears after a gap", e.G)
			}
			seenG[e.G] = true
		}
		prevG = e.G

		// V10 -- line within source bounds.
		if e.Ln != 0 && srcLines > 0 && e.Ln > srcLines {
			add("V10", SevWarning, i, "ln %d is past the end of meta.source.text (%d lines)", e.Ln, srcLines)
		}

		switch e.T {
		case Init:
			if e.S == "" {
				add("V2", SevError, i, "init has no structure name")
				continue
			}
			if _, dup := decls[e.S]; dup {
				add("V2", SevError, i, "structure %q initialised twice", e.S)
				continue
			}
			if e.Kind.IsNodeKind() && e.Schema == nil {
				add("V3", SevError, i, "structure %q is kind %q but declares no schema", e.S, e.Kind)
			}
			if e.Kind == KindGrid && len(e.Dims) != 2 {
				add("V3", SevError, i, "grid %q needs dims [rows,cols], got %v", e.S, e.Dims)
			}
			if e.Kind == KindArray && len(e.Dims) != 1 {
				add("V3", SevError, i, "array %q needs dims [n], got %v", e.S, e.Dims)
			}
			decls[e.S] = e
			aux[e.S] = e.Aux

		case Set:
			decl, ok := decls[e.S]
			if !ok {
				// V2
				add("V2", SevError, i, "set writes to %q, which was never initialised", e.S)
				continue
			}
			// V3 -- address well-formed for the kind.
			if msg := checkAddr(decl, e.At); msg != "" {
				add("V3", SevError, i, "%s", msg)
			}
			// V12 -- reserved node id prefix.
			if decl.Kind.IsNodeKind() && len(e.At) > 0 && !e.At[0].IsIdx {
				head := e.At[0].S
				if strings.HasPrefix(head, "$") && head != NSRefs && head != NSEdges {
					add("V12", SevError, i, "node id %q may not begin with '$'", head)
				}
			}
			// V8 -- detail levels only on aux structures.
			if e.Lvl > 0 && !aux[e.S] {
				add("V8", SevError, i, "lvl %d on %q, which is not declared aux "+
					"(filtering it would corrupt replay -- see ADR 0016)", e.Lvl, e.S)
			}
			// V4 -- from matches current state.
			if cur := state.Get(e.S, e.At); !Equal(cur, e.From) {
				add("V4", SevError, i, "from is %s but %s currently holds %s",
					Canon(e.From), e.At.KeyWith(e.S), Canon(cur))
			}
			// V11 -- redundant no-op write.
			//
			// Only on a REPEAT write. The first write to a cell that happens to
			// equal the fill value is not a bug -- it is a computed cell whose
			// answer is the fill, and it is exactly the case the grid renderer
			// distinguishes as "computed 0" versus "still 0". Flagging it would
			// warn on almost every DP table, which trains you to ignore V11.
			key := e.At.KeyWith(e.S)
			if Equal(e.From, e.To) && written[key] {
				add("V11", SevWarning, i, "%s is written again with the value it already has (%s)",
					key, Canon(e.To))
			}
			written[key] = true

		case Call:
			depth++

		case Ret:
			depth--
			if depth < 0 {
				add("V5", SevError, i, "ret with no open call")
				depth = 0
			}
		}

		// V9 -- deps refer to live structures.
		for _, d := range e.Deps {
			ds, ok := decls[d.S]
			if !ok {
				add("V9", SevError, i, "dep refers to %q, which does not exist yet", d.S)
				continue
			}
			// An aux structure is hidden at detail level 0. If a level-0 event's
			// explanation cites one, the user reads "where mid was 4" about
			// something not on screen.
			//
			// This is a WARNING, not an error, and the distinction is worth
			// stating because the first draft of this rule got it wrong and
			// rejected binary search -- an algorithm the design specifically
			// exists to support.
			//
			// Filtering is sound for a sharper reason than "aux structures are
			// never read": every `set` carries its full `to` value, never a
			// delta, so dropping event X cannot change the value any other event
			// writes. The ONLY casualty of filtering is X's own structure going
			// stale -- which is harmless precisely because `aux` is what makes
			// it hidden. Aux-reads-aux (mid = (lo+hi)/2) is therefore fine: the
			// whole cluster is filtered together.
			if aux[d.S] && e.Lvl == 0 && !aux[e.S] {
				add("V8", SevWarning, i, "explanation cites aux structure %q, "+
					"which is hidden at detail level 0", d.S)
			}
			if msg := checkAddr(ds, d.At); msg != "" {
				add("V9", SevError, i, "dep address: %s", msg)
			}
		}

		if err := state.ApplyForward(i, e, nil); err != nil {
			add("V2", SevError, i, "%v", err)
		}
	}

	// V5 -- unbalanced at the end. Legal only for a truncated trace.
	if depth > 0 && !t.Meta.Truncated {
		diags = append(diags, Diag{"V5", SevError, -1,
			fmt.Sprintf("%d call(s) never returned in a trace that is not marked truncated", depth)})
	}

	// V6 -- full backward replay returns to the initial state.
	for i := len(t.Events) - 1; i >= 0; i-- {
		if err := state.ApplyBackward(i, &t.Events[i], nil); err != nil {
			diags = append(diags, Diag{"V6", SevError, i, err.Error()})
			break
		}
	}
	if h := state.Hash(); h != initialHash {
		diags = append(diags, Diag{"V6", SevError, -1,
			"state after a full forward-then-backward replay differs from the initial state"})
	}

	// V13 -- view hints name real structures.
	for _, v := range t.Meta.Views {
		if v.S == "$calls" {
			continue
		}
		if _, ok := decls[v.S]; !ok {
			diags = append(diags, Diag{"V13", SevWarning, -1,
				fmt.Sprintf("view %q names structure %q, which the trace never creates", v.Family, v.S)})
		}
	}

	// V14 -- byte cap.
	if b, err := Encode(t); err == nil && len(b) > DefaultMaxBytes {
		diags = append(diags, Diag{"V14", SevWarning, -1,
			fmt.Sprintf("serialized trace is %d bytes, over the %d byte soft cap", len(b), DefaultMaxBytes)})
	}
	return diags
}

// DefaultMaxBytes is the soft serialized-size cap. Producers enforce it; the
// validator only warns, because a large trace that arrived is still playable.
const DefaultMaxBytes = 8 << 20

func checkAddr(decl *Event, at Path) string {
	switch decl.Kind {
	case KindScalar:
		if len(at) != 0 {
			return fmt.Sprintf("scalar %q is addressed by [], got %v", decl.S, at.Join())
		}
	case KindArray:
		if len(at) != 1 || !at[0].IsIdx {
			return fmt.Sprintf("array %q needs a single integer index, got %q", decl.S, at.Join())
		}
		if len(decl.Dims) == 1 && (at[0].I < 0 || at[0].I >= decl.Dims[0]) {
			return fmt.Sprintf("index %d is outside array %q of length %d", at[0].I, decl.S, decl.Dims[0])
		}
	case KindGrid:
		if len(at) != 2 || !at[0].IsIdx || !at[1].IsIdx {
			return fmt.Sprintf("grid %q needs two integer indices, got %q", decl.S, at.Join())
		}
		if len(decl.Dims) == 2 {
			if at[0].I < 0 || at[0].I >= decl.Dims[0] || at[1].I < 0 || at[1].I >= decl.Dims[1] {
				return fmt.Sprintf("(%d,%d) is outside grid %q of size %dx%d",
					at[0].I, at[1].I, decl.S, decl.Dims[0], decl.Dims[1])
			}
		}
	case KindMap:
		if len(at) == 0 {
			return fmt.Sprintf("map %q needs at least one key segment", decl.S)
		}
	case KindNodes, KindGraph:
		if len(at) == 0 {
			return fmt.Sprintf("%q needs at least a node id", decl.S)
		}
		if at[0].IsIdx {
			return fmt.Sprintf("%q is addressed by node id, got integer %d", decl.S, at[0].I)
		}
	}
	return ""
}

File: web/src/lib/validate.js
// @ts-check
/**
 * The trust boundary.
 *
 * Every trace is validated here before it reaches the player -- including ones
 * that arrive from the network or from a file the user dragged in. This is the
 * JavaScript twin of trace.Validate in Go, implementing the same V1..V14, and
 * it is the single place where "is this data shaped the way we think" is asked.
 *
 * Concentrating the check here is also what makes writing the rest of the
 * frontend in plain JavaScript defensible (ADR 0009): past this function the
 * shape is known-good, so the remaining risk is renderers misreading VALID
 * data -- a far smaller surface than "anything could be anything".
 *
 * It must never throw. A malformed trace produces diagnostics, not an
 * exception, because the UI renders those diagnostics.
 */

import { State } from '../player/state.js';
import { addrKey, canon, deepEqual } from './value.js';

export const SUPPORTED_MIN = 1;
export const SUPPORTED_MAX = 1;

/**
 * @typedef {{check:string, severity:'error'|'warning', event:number, message:string}} Diag
 */

const NODE_KINDS = new Set(['nodes', 'graph']);

/**
 * @param {*} trace
 * @returns {Diag[]}
 */
export function validate(trace) {
  /** @type {Diag[]} */
  const out = [];
  const add = (check, severity, event, message) => out.push({ check, severity, event, message });

  try {
    if (!trace || typeof trace !== 'object') {
      add('V1', 'error', -1, 'not a trace object');
      return out;
    }
    if (typeof trace.v !== 'number' || trace.v < SUPPORTED_MIN || trace.v > SUPPORTED_MAX) {
      add('V1', 'error', -1,
        `this trace is version ${trace.v}; this build reads versions ${SUPPORTED_MIN}–${SUPPORTED_MAX}`);
      return out;
    }
    if (!Array.isArray(trace.events)) {
      add('V1', 'error', -1, 'trace.events is missing or not an array');
      return out;
    }

    const decls = new Map();
    const aux = new Set();
    const written = new Set();
    const state = new State(trace);
    const initialHash = state.hash();

    let depth = 0;
    let prevG = 0;
    const seenG = new Set();
    const srcLines = trace.meta?.source?.text ? trace.meta.source.text.split('\n').length : 0;

    for (let i = 0; i < trace.events.length; i++) {
      const e = trace.events[i];
      if (!e || typeof e !== 'object') {
        add('V1', 'error', i, 'event is not an object');
        continue;
      }

      // V7 -- group ids form contiguous runs.
      const g = e.g ?? 0;
      if (g !== 0) {
        if (g !== prevG && seenG.has(g)) add('V7', 'error', i, `group id ${g} reappears after a gap`);
        seenG.add(g);
      }
      prevG = g;

      // V10 -- line within source bounds.
      if (e.ln && srcLines && e.ln > srcLines) {
        add('V10', 'warning', i, `ln ${e.ln} is past the end of meta.source.text (${srcLines} lines)`);
      }

      switch (e.t) {
        case 'init': {
          if (!e.s) { add('V2', 'error', i, 'init has no structure name'); continue; }
          if (decls.has(e.s)) { add('V2', 'error', i, `structure "${e.s}" initialised twice`); continue; }
          if (NODE_KINDS.has(e.kind) && !e.schema) {
            add('V3', 'error', i, `structure "${e.s}" is kind "${e.kind}" but declares no schema`);
          }
          if (e.kind === 'grid' && (!Array.isArray(e.dims) || e.dims.length !== 2)) {
            add('V3', 'error', i, `grid "${e.s}" needs dims [rows, cols]`);
          }
          if (e.kind === 'array' && (!Array.isArray(e.dims) || e.dims.length !== 1)) {
            add('V3', 'error', i, `array "${e.s}" needs dims [n]`);
          }
          decls.set(e.s, e);
          if (e.aux) aux.add(e.s);
          break;
        }
        case 'set': {
          const decl = decls.get(e.s);
          if (!decl) {
            add('V2', 'error', i, `set writes to "${e.s}", which was never initialised`);
            continue;
          }
          const at = e.at ?? [];
          const msg = checkAddr(decl, at);
          if (msg) add('V3', 'error', i, msg);

          // V12 -- reserved node id prefix.
          if (NODE_KINDS.has(decl.kind) && typeof at[0] === 'string' &&
              at[0].startsWith('$') && at[0] !== '$refs' && at[0] !== '$edges') {
            add('V12', 'error', i, `node id "${at[0]}" may not begin with '$'`);
          }
          // V8 -- detail levels only on aux structures.
          if ((e.lvl ?? 0) > 0 && !aux.has(e.s)) {
            add('V8', 'error', i,
              `lvl ${e.lvl} on "${e.s}", which is not declared aux — filtering it would corrupt replay`);
          }
          // V4 -- from matches current state. With V6 this is a complete proof
          // that the trace is a valid reversible delta log.
          const cur = state.get(e.s, at);
          if (!deepEqual(cur, e.from ?? null)) {
            add('V4', 'error', i,
              `from is ${canon(e.from ?? null)} but ${addrKey(e.s, at)} currently holds ${canon(cur)}`);
          }
          // V11 -- redundant no-op write, on a REPEAT only. A first write whose
          // value happens to equal the fill is a computed cell, not a bug.
          const key = addrKey(e.s, at);
          if (deepEqual(e.from ?? null, e.to ?? null) && written.has(key)) {
            add('V11', 'warning', i, `${key} is written again with the value it already has`);
          }
          written.add(key);
          break;
        }
        case 'call':
          depth++;
          break;
        case 'ret':
          depth--;
          if (depth < 0) { add('V5', 'error', i, 'ret with no open call'); depth = 0; }
          break;
        default:
          add('V1', 'error', i, `unknown event type "${e.t}"`);
      }

      // V9 -- deps refer to live structures.
      for (const d of e.deps ?? []) {
        const ds = decls.get(d.s);
        if (!ds) { add('V9', 'error', i, `dep refers to "${d.s}", which does not exist yet`); continue; }
        if (aux.has(d.s) && (e.lvl ?? 0) === 0 && !aux.has(e.s)) {
          add('V8', 'warning', i,
            `explanation cites aux structure "${d.s}", which is hidden at detail level 0`);
        }
        const m = checkAddr(ds, d.at ?? []);
        if (m) add('V9', 'error', i, `dep address: ${m}`);
      }

      state.forward(i, e, null);
    }

    if (depth > 0 && !trace.meta?.truncated) {
      add('V5', 'error', -1, `${depth} call(s) never returned in a trace that is not marked truncated`);
    }

    // V6 -- full backward replay returns to the initial state.
    for (let i = trace.events.length - 1; i >= 0; i--) state.backward(i, trace.events[i], null);
    if (state.hash() !== initialHash) {
      add('V6', 'error', -1,
        'state after a full forward-then-backward replay differs from the initial state');
    }

    // V13 -- view hints name real structures.
    for (const v of trace.meta?.views ?? []) {
      if (v.s === '$calls') continue;
      if (!decls.has(v.s)) {
        add('V13', 'warning', -1,
          `view "${v.family}" names structure "${v.s}", which the trace never creates`);
      }
    }
  } catch (err) {
    add('V0', 'error', -1, `validator threw: ${err && err.message} (this is a bug in Orrery)`);
  }
  return out;
}

/** @param {Diag[]} ds */
export function hasErrors(ds) {
  return ds.some((d) => d.severity === 'error');
}

function checkAddr(decl, at) {
  switch (decl.kind) {
    case 'scalar':
      if (at.length !== 0) return `scalar "${decl.s}" is addressed by [], got ${at.join('/')}`;
      break;
    case 'array':
      if (at.length !== 1 || typeof at[0] !== 'number') {
        return `array "${decl.s}" needs a single integer index, got "${at.join('/')}"`;
      }
      if (Array.isArray(decl.dims) && (at[0] < 0 || at[0] >= decl.dims[0])) {
        return `index ${at[0]} is outside array "${decl.s}" of length ${decl.dims[0]}`;
      }
      break;
    case 'grid':
      if (at.length !== 2 || typeof at[0] !== 'number' || typeof at[1] !== 'number') {
        return `grid "${decl.s}" needs two integer indices, got "${at.join('/')}"`;
      }
      if (Array.isArray(decl.dims) &&
          (at[0] < 0 || at[0] >= decl.dims[0] || at[1] < 0 || at[1] >= decl.dims[1])) {
        return `(${at[0]},${at[1]}) is outside grid "${decl.s}" of size ${decl.dims[0]}×${decl.dims[1]}`;
      }
      break;
    case 'map':
      if (at.length === 0) return `map "${decl.s}" needs at least one key segment`;
      break;
    case 'nodes':
    case 'graph':
      if (at.length === 0) return `"${decl.s}" needs at least a node id`;
      if (typeof at[0] === 'number') return `"${decl.s}" is addressed by node id, got integer ${at[0]}`;
      break;
  }
  return '';
}
